src/sentry/event_manager.py

```diff
@@ -1891,6 +1891,55 @@ def _create_group(project: Project, event: Event, **group_creation_kwargs: Any)
     )
 
 
+def _is_stuck_counter_error(err: Exception, project: Project, short_id: int) -> bool:
+    """Decide if this is `UniqueViolation` error on the `Group` table's project and short id values."""
+
+    error_message = err.args[0]
+
+    for substring in [
+        f"Key (project_id, short_id)=({project.id}, {short_id}) already exists.",
+        'duplicate key value violates unique constraint "sentry_groupedmessage_project_id_short_id',
+    ]:
+        if substring in error_message:
+            return True
+
+    return False
+
+
+def _handle_stuck_project_counter(project: Project, current_short_id: int) -> int:
+    """
+    Sometimes, for reasons unknown, a project's `Counter` value falls behind its latest group id.
+    When that happens, that incorrect counter value leads us to try to create groups with ids which
+    are already taken.
+
+    This handles that case by updating the counter's value to the latest group id, and then returns
+    the new value.
+    """
+    new_short_id = current_short_id
+
+    # Ordinarily running max on this many rows would be prohibitively expensive, but a) this is
+    # a very rare case (< 20 ever that we know of), and b) project and short id are indexed
+    # together in order to enforce the unique constraint which got us here in the first place,
+    # so it's faster than it otherwise might be.
+    max_short_id_for_project = Group.objects.filter(project_id=project.id).aggregate(
+        Max("short_id")
+    )["short_id__max"]
```

Comment:
To some extent I wonder if we should just skip the query and arbitrarily increment by 5 or something, but I guess this is cheap/rare enough that it's not a big problem